### PR TITLE
Add developer guidelines

### PR DESCRIPTION
**Pull request by nolmoonen**
Adds the first version of developer guidelines to guide contributors.

### CODE REVIEW

**Review by lpaoletti — APPROVED** [review]
Took a quick look at the deverloper.rst file. 
Looks good.

**Review by RobsonRLemos — APPROVED** [review]
The document is well structured and presents the level of information needed to developers. Also, the code examples provide a better understanding of the rocPRIM’s design architecture.

File: docs/concepts/developer.rst [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+To facilitate a dispatching mechanism supporting the above requirements, several standardized structures need to be defined for each algorithm, which is outlined in this section. These structures depend on a generalized dispatching mechanism.
+
+The algorithm's configuration struct is defined in ``rocprim/device/detail/device_config_helper.hpp``. The reason for putting all configurations in one file is to make the configuration templates simpler (generating configurations is explained in the contributor's guideline). The tuning config has the name ``ALGO_config``, and no members (unless for backward-compatibility reasons), only template parameters. 
+
+The config struct derives from a non-public parameter struct holding the actual parameters. This separation between structs is done facilitate change without breaking public API.
```
Comment (stanleytsang-amd) [inline docs/concepts/developer.rst]: Minor typo: This separation between structs is done **to**

**Review by stanleytsang-amd — CHANGES_REQUESTED** [review]
Aside from minor typo, everything looks good to me.  Great starting guide.

File: docs/concepts/developer.rst (hunk 1 above)
Comment (nolmoonen) [reply docs/concepts/developer.rst]: Fixed.

**Review by stanleytsang-amd — APPROVED** [review]

**Pull request merged**